internal/provider/resource_bytes_test.go

```diff
@@ -0,0 +1,643 @@
+package provider
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccResourceBytes(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		ProtoV5ProviderFactories: protoV5ProviderFactories(),
+		Steps: []resource.TestStep{
+			{
+				Config: `resource "random_bytes" "basic" {
+							length = 32
+						}`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestMatchResourceAttr("random_bytes.basic", "result_base64", regexp.MustCompile(`^[A-Za-z/+\d]{43}=$`)),
+					resource.TestMatchResourceAttr("random_bytes.basic", "result_hex", regexp.MustCompile(`^[a-f\d]{64}$`)),
+					resource.TestCheckResourceAttr("random_bytes.basic", "length", "32"),
+				),
+			},
+			{
+				// Usage of ImportStateIdFunc is required as the value passed to the `terraform import` command needs
+				// to be the bytes encoded with base64, as the bytes resource sets ID to "none"
+				ImportStateIdFunc: func(s *terraform.State) (string, error) {
+					id := "random_bytes.basic"
+					rs, ok := s.RootModule().Resources[id]
+					if !ok {
+						return "", fmt.Errorf("not found: %s", id)
+					}
+					if rs.Primary.ID == "" {
+						return "", fmt.Errorf("no ID is set")
+					}
+
+					return rs.Primary.Attributes["result_base64"], nil
+				},
+				ResourceName:      "random_bytes.basic",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccResourceBytes_ImportWithoutKeepersThenUpdateShouldNotTriggerChange(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		ProtoV5ProviderFactories: protoV5ProviderFactories(),
+		Steps: []resource.TestStep{
+			{
+				ImportState:        true,
+				ImportStateId:      "hkvbcU5f8qGysTFhkI4gzf3yRWC1jXW3aRLCNQFOtNw=",
+				ImportStatePersist: true,
+				ResourceName:       "random_bytes.basic",
+				Config: `resource "random_bytes" "basic" {
+							length = 32
+						}`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("random_bytes.basic", "result_base64", "hkvbcU5f8qGysTFhkI4gzf3yRWC1jXW3aRLCNQFOtNw="),
+					resource.TestCheckResourceAttr("random_bytes.basic", "result_hex", "864bdb714e5ff2a1b2b13161908e20cdfdf24560b58d75b76912c235014eb4dc"),
+					resource.TestCheckResourceAttr("random_bytes.basic", "length", "32"),
+				),
+			},
+			{
+				ResourceName: "random_bytes.basic",
+				Config: `resource "random_bytes" "basic" {
+							length = 32
+						}`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("random_bytes.basic", "result_base64", "hkvbcU5f8qGysTFhkI4gzf3yRWC1jXW3aRLCNQFOtNw="),
+				),
```

Comment:
Using `types.MapNull(types.StringType)` rather than `types.MapValueMust(types.StringType, nil)` prevents an in-place update if the `random_bytes` resource is imported and then `terraform apply` is run.

Could you amend the test as follows:

```suggestion
				PlanOnly: true,
```

If you try running this test whilst using `types.MapValueMust(types.StringType, nil)` you can see that it fails with the following error:

```          # random_bytes.basic will be updated in-place
          ~ resource "random_bytes" "basic" {
                id            = "none"
              - keepers       = {} -> null
                # (3 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
```